README.md
# ALICIA
Augmented Lagrangian Implicit Constrained Inverse Analysis

## Description
ALICIA is a code for calculating heat flux density at tokamak PFCs. It is mainly developed for increased accuracy in the measurement of the SOL normal heat flux density at irregular divertor tiles using tempertures acquired by fast IR cameras.

ALICIA uses MkniX library for generating and solving the 2D numerical FEM models, and is distributed under the GPL v2 license (see License section below).

## Installation

ALICIA uses the cmake build system for cross-platform compatibility (version 2.8 or greater).

Before building ALICIA, the source code and shared library of MkniX shall be available locally. Please use Mknix devel branch.

Build ALICIA using (from the ALICIA root directory):

```
cmake -B<build directory> -H. -DCMAKE_BUILD_TYPE=<Debug|Release> -DMKNIX_LINK_DIR=<path/to/dir/containing-libmknix.so> -DMKNIX_SOURCE_DIR=<path/to/dir/containing/MkniX>
```

The use of cmake-gui is recommended. Alternatively, an example of the cmake one-liner command is provided (note that paths might differ for each user):

```
cmake -B../ALICIA-build -H. -DCMAKE_BUILD_TYPE=Release -DMKNIX_LINK_DIR=../MkniX-devel-build/src/ -DMKNIX_SOURCE_DIR=../MkniX
```

## Examples

ALICIA is provided with some examples within the folder "ALICIA-examples" so it can be tested in the deployment platform.

To list all example directories:
```
cd <build directory>/ALICIA-examples
ls
```
## Usage

Once navigated to an specific example or working directory, ALICIA can be run using two command line arguments: (1) the MkniX input file, and (2) the IR temperatures file.

e.g.
```
cd <build directory>/ALICIA-examples/90271-TPT6-augmented-5deg-alpha1E3-res1E1/
../../src/alicia input_tile6_fem.mknix temps_IR.txt 
```

## License

Copyright (C) 2007-2017 by Daniel Iglesias

Copyright (c) 2013-2018, European Atomic Energy Community (EURATOM)

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.



CHANGELOG.txt
Project Changelog
=================

Release 0.9.0 (10 Apr 2019)
--------------------------

API Changes:
* <<<List of changes>>>

Fixed:
* <<<List of changes>>>

New:
* <<<List of changes>>>

Deprecated:
* <<<List of changes>>>

Removed:
* <<<List of changes>>>


scripts/compute_power.py
#!/usr/bin/env python

#   Copyright (C) 2007-2017 by Daniel Iglesias                            
#   Copyright (c) 2013-2018, European Atomic Energy Community (EURATOM)   
#   https://github.com/daniel-iglesias/alicia                             
#                                                                         
#   This program is free software; you can redistribute it and/or modify  
#   it under the terms of the GNU Lesser General Public License as        
#   published by the Free Software Foundation; either version 2 of the    
#   License, or (at your option) any later version.                       
#                                                                         
#   This program is distributed in the hope that it will be useful,       
#   but WITHOUT ANY WARRANTY; without even the implied warranty of        
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         
#   GNU General Public License for more details.                          
#                                                                         
#   You should have received a copy of the GNU Lesser General Public      
#   License along with this program; if not, write to the                 
#   Free Software Foundation, Inc.,                                       
#   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             

import sys
import math
import numpy as np
from scipy.interpolate import interp1d

# Tile-dependent parameters
# read the radius from the centre of the machine
radius_in = np.loadtxt("stackA_radius.txt")

# Inputs from ALICIA
data = np.loadtxt("flux_interface.txt")
profile = data[0,1:]
time = data[1:,0]
flux = data[1:,1:]

# Inputs from FLUSH
# TODO

# Calculation of TWF
# TODO

# Calculate node_weights using profile info
node_weights = np.empty(profile.shape)
for i in range(profile.size-1) :
    node_weights[i] += (profile[i+1]-profile[i]) / 2.
    node_weights[i+1] += (profile[i+1]-profile[i]) / 2.

# Power integrated poloidally
power = np.dot(flux, node_weights)

time_power = np.column_stack(
	(time,
	power))
np.savetxt("power_local.txt",time_power)

# Total power
interp_r_fun = interp1d(radius_in[0],radius_in[1])
radius = interp_r_fun(profile)

# Calculate node_weights using radius info
node_par_weights = np.empty(profile.shape)
for i in range(radius.size-1) :
    # weight = 2*pi*r * delta_r
    node_par_weights[i] += 2*math.pi*radius[i] * (radius[i+1]-radius[i]) / 2.
    node_par_weights[i+1] += 2*math.pi*radius[i] * (radius[i+1]-radius[i]) / 2.

power_total = np.dot(flux*TWF, node_par_weights)

time_power = np.column_stack(
	(time,
	power_total))
np.savetxt("power_total.txt",time_power)

# And the energy evolution
delta_t = (time[-1]-time[0])/(time.size-1)
energy = np.zeros(power_total.shape)

for i in range(power_total.size-1) :
    # Trapezoidal rule
    energy[i+1] = energy[i] + delta_t * (power_total[i] + power_total[i+1])/2.
    
time_energy = np.column_stack(
	(time,
	energy
	))
np.savetxt("energy.txt",time_energy)



code-templates/template.c
/***************************************************************************
 *   Copyright (C) 2007-2017 by Daniel Iglesias                            *
 *   Copyright (c) 2013-2018, European Atomic Energy Community (EURATOM)   *
 *   https://github.com/daniel-iglesias/alicia                             *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Lesser General Public License as        *
 *   published by the Free Software Foundation; either version 2 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU Lesser General Public      *
 *   License along with this program; if not, write to the                 *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

code-templates/template.py
#   Copyright (C) 2007-2017 by Daniel Iglesias                            
#   Copyright (c) 2013-2018, European Atomic Energy Community (EURATOM)   
#   https://github.com/daniel-iglesias/alicia                             
#                                                                         
#   This program is free software; you can redistribute it and/or modify  
#   it under the terms of the GNU Lesser General Public License as        
#   published by the Free Software Foundation; either version 2 of the    
#   License, or (at your option) any later version.                       
#                                                                         
#   This program is distributed in the hope that it will be useful,       
#   but WITHOUT ANY WARRANTY; without even the implied warranty of        
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         
#   GNU General Public License for more details.                          
#                                                                         
#   You should have received a copy of the GNU Lesser General Public      
#   License along with this program; if not, write to the                 
#   Free Software Foundation, Inc.,                                       
#   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             


scripts/compute_max_power_density.py
#!/usr/bin/env python

#   Copyright (C) 2007-2017 by Daniel Iglesias                            
#   Copyright (c) 2013-2018, European Atomic Energy Community (EURATOM)   
#   https://github.com/daniel-iglesias/alicia                             
#                                                                         
#   This program is free software; you can redistribute it and/or modify  
#   it under the terms of the GNU Lesser General Public License as        
#   published by the Free Software Foundation; either version 2 of the    
#   License, or (at your option) any later version.                       
#                                                                         
#   This program is distributed in the hope that it will be useful,       
#   but WITHOUT ANY WARRANTY; without even the implied warranty of        
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         
#   GNU General Public License for more details.                          
#                                                                         
#   You should have received a copy of the GNU Lesser General Public      
#   License along with this program; if not, write to the                 
#   Free Software Foundation, Inc.,                                       
#   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             

import sys
import math
import numpy as np
from scipy.interpolate import interp1d

data = np.loadtxt("flux_interface.txt")
profile = data[0,1:]
time = data[1:,0]
flux = data[1:,1:]

# The moving avg. is calculated for the original fluxes:
flux_max = flux.max(axis=1)
time_flux_max = np.column_stack(
	(time,
	flux_max))

np.savetxt("flux_interface_max.txt", time_flux_max)




src/ALICIA.cpp
/***************************************************************************
 *   Copyright (C) 2007-2017 by Daniel Iglesias                            *
 *   Copyright (c) 2013-2018, European Atomic Energy Community (EURATOM)   *
 *   https://github.com/daniel-iglesias/alicia                             *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Lesser General Public License as        *
 *   published by the Free Software Foundation; either version 2 of the    *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU Lesser General Public      *
 *   License along with this program; if not, write to the                 *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <cstdlib>
//#include <unistd.h>

#include "simulation.h"
#include "node.h"
#include "common.h"

using namespace std;
using namespace mknix;

int main(int argc, char *argv[])
{
    lmx::setMatrixType( 3 );
    lmx::setLinSolverType( 2 );

    if (argc >= 3 && argc < 5) {
        //     cout << "EXECUTING: $ alicia " << argv[1] << endl;
        //     system("echo -n '1. Current Directory is '; pwd");
        mknix::Simulation mySimulation;
        mySimulation.setOutputFilesDetail( 1 ); // 0 - nothing, 1 - timing, 2 - All
        mySimulation.inputFromFile(argv[1]);

        std::string systemName = "tiles";
        std::ifstream ifile(argv[2]);

        boxFIR box(1); //If this is 1, then no filtering happens, use bigger ints for more smoothing

        std::ofstream interface_file ("interface_coordinates.txt");
        int numberOfInterfaceNodes = mySimulation.getInterfaceNumberOfNodes(systemName);
        vector<double> interface_coordinates;
        // Fill x coordinates for constrained nodes...
        for(int i=0; i<numberOfInterfaceNodes; ++i){
            interface_coordinates.push_back(mySimulation.getInterfaceNode(systemName, i)->getX());
            interface_file << interface_coordinates.back() << "\t";
        }
        interface_file << endl;
        cout << "NODES: " << numberOfInterfaceNodes << endl;
        cout << interface_coordinates[0] << " ... " << interface_coordinates.back() << endl;

        const std::vector<std::vector<double> > vectors = read_lines(ifile) ;

        double t_0 = vectors[2][0];
        double time = t_0;
        double t_n = vectors.back()[0];
        int iterations = (vectors.size()-3);
        double delta_t = (t_n - t_0)/iterations;
        int infoEveryNthIteration = iterations/100+1; // Every 1 %
        cout << "t_0 = " << t_0 << endl;
        cout << "t_n = " << t_n << endl;
        cout << "iterations = " << iterations << endl;
        cout << "delta_t = " << delta_t << endl;
        if (argc >= 4){
            iterations =2; //doing dry runs to check time parameters
            if (argv[3] == "dry") return EXIT_SUCCESS;
        }

        // initializing coordinate map:
        map<double, double> input_load_step;
        int input_points_num = vectors[0].size();
        for(int i=0; i<input_points_num; ++i){
            input_load_step[ vectors[0][i] ]= 0.;
            interface_file << vectors[0][i] << "\t";
        }
        interface_file << endl;
//         mySimulation.setInitialTemperatures(0); // Validation & tortures
//        mySimulation.setInitialTemperatures(200); // 3D tortures (3 & 4)
        mySimulation.setInitialTemperatures(150); // Reference in-vessel value
//         mySimulation.setInitialTemperatures(23); // Reference MAST-U
        mySimulation.init(2); // Output Vervosity=0

        std::ofstream temps_filtered ("temps_filtered.txt");
        std::ofstream temps_interface ("temps_interface.txt");
        std::ofstream flux_interface ("flux_interface.txt");
        std::ofstream flux_max_min ("flux_interface_max_min.txt");

        // Output 2D files m-type matrix preparation: size of time vector in first element, x-coordinates in the rest of the 1st row.
        temps_interface << iterations << "\t";
        flux_interface << iterations << "\t";
        for(int i=0; i<numberOfInterfaceNodes; ++i){
            temps_interface << mySimulation.getInterfaceNode(systemName, i)->getX() << "\t";
            flux_interface << mySimulation.getInterfaceNode(systemName, i)->getX() << "\t";
        }
        temps_interface << endl;
        flux_interface << endl;

        //         double factor(1./0.00387); //width of the element
        //         factor*=0.836695; // power equilibrium between flat and 3D shadowed as measured in CAD; nominal position
        //         factor*=0.85; // Reduction of thermal capacity due to the gap (less material to heat)
        //         factor*=0.9; // Effect of the change in section of the lamella; the mid part of the lamella is thinner.

        for(int i=0; i<iterations; ++i){
            vector<double> filtered_vector ( vectors[i+2] );
            filtered_vector.erase(filtered_vector.begin()); // Remove time information
            vector<double> interface_fluxes( interface_coordinates.size() );

             box.filter(filtered_vector);
            //             cout << "input_points_num = " << input_points_num << endl;
            //             cout << "input_load_step.size() = " << input_load_step.size() << endl;
            //             cout << "filtered_vector.size() = " << filtered_vector.size() << endl;
            for(int j=0; j<input_points_num; ++j){
                input_load_step.at( vectors[0][j] ) = filtered_vector[j];
                temps_filtered << input_load_step.at( vectors[0][j] ) << '\t';
            }
            temps_filtered << '\n';

            temps_interface << time << '\t';
            flux_interface << time << '\t';

            for(int j=0; j<numberOfInterfaceNodes; ++j){
                mySimulation.getInterfaceNode(systemName,j)->setqt( interpolate1D( interface_coordinates[j], input_load_step ) );
                temps_interface << mySimulation.getInterfaceNode(systemName,j)->getqt() << '\t';
            }
            temps_interface << '\n';
            mySimulation.solveStep();
            // First node:
            {
                std::string constraintName( "INT_" + std::to_string(0) );
                interface_fluxes.push_back( mySimulation.getConstraintOutput(constraintName, systemName, 0)
                / ( std::abs(interface_coordinates[1]-interface_coordinates[0]) / 2. ) ) ; // scaled with the global Qperp factor and border condition
                flux_interface << interface_fluxes.back() << '\t';
            }

            for(int j=1; j<numberOfInterfaceNodes-1; ++j){
                std::string constraintName( "INT_" + std::to_string(j) );
                //                 cout << "Getting constraint data from " << systemName << "." << constraintName << endl;
                interface_fluxes.push_back( mySimulation.getConstraintOutput(constraintName, systemName, 0)
                / ( std::abs(interface_coordinates[j+1]-interface_coordinates[j-1]) / 2.) ) ;
                flux_interface << interface_fluxes.back() << '\t';
            }
            // Last node:
            {
                std::string constraintName( "INT_" + std::to_string(numberOfInterfaceNodes-1) );
                interface_fluxes.push_back( mySimulation.getConstraintOutput(constraintName, systemName, 0)
                / ( std::abs(interface_coordinates.back()-interface_coordinates[numberOfInterfaceNodes-2]) / 2. ) ) ; // scaled with the global Qperp factor and border condition
                flux_interface << interface_fluxes.back() << '\t';
            }

            flux_interface << '\n';
            flux_max_min << time << '\t'
            << *max_element(std::begin(interface_fluxes), std::end(interface_fluxes))  << '\t'
            << *min_element(std::begin(interface_fluxes), std::end(interface_fluxes))
            << '\n';
            time += delta_t;
            if( i % infoEveryNthIteration == 0 ){
                cout << int(double(i)/iterations*100) << " % Complete," << "\t TIME = " << time << ",\t STEP = " << i << endl;
            }
        }
        mySimulation.endSimulation();

    } else if(argc != 2) {
        cout<< "Need two parameters: 1. mknix input file, and 2. Temperature map."<<endl
        << "Usage: $./alicia \"mknix_file_name\" \"IR_temps_file\""<<endl;
    }
    return EXIT_SUCCESS;
}
